finality-aleph/src/new_network/connection_manager/discovery.rs

```diff
@@ -0,0 +1,690 @@
+        use DiscoveryMessage::*;
+        match message {
+            AuthenticationBroadcast(authentication) => {
+                self.handle_broadcast(authentication, handler)
+            }
+            Request(node_ids, authentication) => {
+                self.handle_request(node_ids, authentication, handler)
+            }
+            Authentications(authentications) => {
+                (self.handle_response(authentications, handler), None)
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{Discovery, DiscoveryMessage};
+    use crate::{
+        crypto::{AuthorityPen, AuthorityVerifier, KeyBox},
+        new_network::{
+            connection_manager::{Authentication, SessionHandler},
+            DataCommand, Multiaddr, Protocol,
+        },
+        NodeIndex, SessionId,
+    };
+    use aleph_primitives::{AuthorityId, KEY_TYPE};
+    use codec::Encode;
+    use sc_network::{
+        multiaddr::Protocol as ScProtocol, Multiaddr as ScMultiaddr, PeerId as ScPeerId,
+    };
+    use sp_keystore::{testing::KeyStore, CryptoStore};
+    use std::{collections::HashSet, iter, net::Ipv4Addr, sync::Arc, thread::sleep, time};
+
+    const NUM_NODES: u8 = 7;
+    const MS_COOLDOWN: u64 = 200;
+
+    fn addresses() -> Vec<Multiaddr> {
+        (0..NUM_NODES)
+            .map(|id| {
+                ScMultiaddr::empty()
+                    .with(ScProtocol::Ip4(Ipv4Addr::new(192, 168, 1, id)))
+                    .with(ScProtocol::Tcp(30333))
+                    .with(ScProtocol::P2p(ScPeerId::random().into()))
+            })
+            .collect()
+    }
+
+    async fn keyboxes() -> Vec<KeyBox> {
+        let num_keyboxes: usize = NUM_NODES.into();
+        let keystore = Arc::new(KeyStore::new());
+        let mut auth_ids = Vec::with_capacity(num_keyboxes);
+        for _ in 0..num_keyboxes {
+            let pk = keystore.ed25519_generate_new(KEY_TYPE, None).await.unwrap();
+            auth_ids.push(AuthorityId::from(pk));
+        }
+        let mut result = Vec::with_capacity(num_keyboxes);
+        for i in 0..num_keyboxes {
+            result.push(KeyBox::new(
+                NodeIndex(i),
+                AuthorityVerifier::new(auth_ids.clone()),
+                AuthorityPen::new(auth_ids[i].clone(), keystore.clone())
+                    .await
+                    .expect("The keys should sign successfully"),
+            ));
+        }
+        result
+    }
+
+    async fn build() -> (Discovery, Vec<SessionHandler>) {
+        let mut handlers = Vec::new();
+        for (keybox, address) in keyboxes().await.into_iter().zip(addresses()) {
+            handlers.push(
+                SessionHandler::new(keybox, SessionId(43), vec![address.into()])
+                    .await
+                    .unwrap(),
+            );
+        }
+        (
+            Discovery::new(time::Duration::from_millis(MS_COOLDOWN)),
+            handlers,
+        )
+    }
+
+    #[tokio::test]
```

Comment:
follow-up to the today slack discussion: is there any testcase in this module, that should be rather run with multithreading? I do not see such, but just to be doubly sure

Reply:
I am assuming all the calls to a specific `Discovery` are performed on one thread, so probably not. On the other hand many of the tests have multiple `Discovery`s, which theoretically could be run on separate threads, only this would be hard to arrange and not really add any value (as they are completely separate entities sharing nothing).